### PR TITLE
Add QueryExpression::newCaseExpr() 

### PR DESCRIPTION
**Pull request by raul338**
Refs #11793

CaseExpression was ~almost~ already fluid, but It was not exposed via Query.  
Now sql cases can mimic real sql cases

Rewrote part of a test, It can be used as example in the cookbook

Happy Easter

### CODE REVIEW

**Comment by raul338** [comment]
AppVeyor fail seems unrelated

File: src/Database/Expression/QueryExpression.php [hunk 1]
```diff
@@ -376,6 +376,7 @@ public function in($field, $values, $type = null)
      * @param array $types associative array of types to be associated with the values
      * passed in $values
      * @return $this
+     * @deprecated 3.6.0 use \Cake\Database\Query::newCaseExpr()
```
Comment (markstory) [inline src/Database/Expression/QueryExpression.php]: Should this also emit a deprecation warning? Or are you planning for the method to last until 5.0?

File: src/Database/Query.php [hunk 2]
```diff
@@ -1746,6 +1747,16 @@ public function newExpr($rawExpression = null)
         return $expression;
     }
 
+    /**
+     * Returns a new CaseExpression object to build SQL Case clauses
+     *
+     * @return \Cake\Database\Expression\CaseExpression
+     */
+    public function newCaseExpr()
```
Comment (markstory) [inline src/Database/Query.php]: Why is this method here and not on QueryExpression like the other expression builders?

Comment (raul338) [reply src/Database/Query.php]: Based on my [original comment](https://github.com/cakephp/cakephp/issues/11793#issuecomment-370888809), I try to make that code working.
`$query->newExpr()->newCaseExpr()` would be more verbose I think

File: src/Database/Expression/QueryExpression.php (hunk 1 above)
Comment (raul338) [reply src/Database/Expression/QueryExpression.php]: Forgot about the warning, I'll add it in a commit

**Comment by raul338** [comment]
If this is good  to go, should I rewrite other `addCase` test to this format?

**Comment by markstory** [comment]
@raul338 We should keep at least one test case for the existing `addCase()` method. You could add another test case if the current builder doesn't have enough coverage.

File: tests/TestCase/Database/QueryTest.php [hunk 3]
```diff
@@ -4221,11 +4276,15 @@ public function testSqlCaseStatement()
             'Not published',
             'None'
         ];
+        $finalExpression = $query->newExpr();
+        $this->withErrorReporting(E_ALL ^ E_USER_DEPRECATED, function () use ($conditions, $values, $finalExpression) {
+            $finalExpression->addCase($conditions, $values);
+        });
```
Comment (raul338) [inline tests/TestCase/Database/QueryTest.php]: I don't know if this is the best way to test for deprecated behavior. Should I change this?

**Comment by lorenzo** [comment]
The new method added to the query object feels out of place. Why is it the only expression type that gets its own method in the query?

**Comment by raul338** [comment]
Right now with this PR this code should work
```php
$publishedCase = $query
    ->newCaseExpr()
    ->add(
        $query->newExpr()->add(['published' => 'Y']),
        1,
        'integer'
    );
```

Would you like this way?
```php
$publishedCase = $query
    ->newExpr()
    ->newCaseExpr()
    ->add(
        $query->newExpr()->add(['published' => 'Y']),
        1,
        'integer'
    );
```
I feel the extra newExpr() a bit odd, it add no meaning

File: tests/TestCase/Database/QueryTest.php (hunk 3 above)
Comment (markstory) [reply tests/TestCase/Database/QueryTest.php]: We have a `$this->deprecated (function () { ... });` helper method for toggling error reporting allowing deprecated methods to be called in tests.

**Comment by markstory** [comment]
The API proposed in #11793 added `when()` and `then()` methods to make defining case expressions easier.  This change gets closer to the previous proposal but doesn't deliver what I saw as the big win which was making when/then cases easier to express.

**Comment by raul338** [comment]
I don't think `when()..then()` would work, what would happend if only one of when/then is called multiple times in a row()?

The original issue is that conditions and values are explicitly separated, instead of being inlined (like resulting sql). In this PR instead of using `when/then`, I just use `add(condition, value, [type])`

Again, see [my original comment](https://github.com/cakephp/cakephp/issues/11793#issuecomment-370888809)

**Comment by markstory** [comment]
> what would happend if only one of when/then is called multiple times in a row?

The resulting SQL would be invalid and the developer would know they made a mistake.

**Comment by raul338** [comment]
Ok, Added when/then functions, a ability to test for simpler column cases.
Maybe needs some more testing, and a bit of rewording in some docblocks

Any failing tests would be fixed tomorrow. Its late here.

Let me know what you think. There is also the issue if move `$query->newCaseExpr()` to something else , perhaps `$query->newExpr()->newCase()` ?

File: src/Database/Expression/CaseExpression.php [hunk 4]
```diff
@@ -52,15 +59,28 @@ class CaseExpression implements ExpressionInterface
     /**
      * Constructs the case expression
      *
-     * @param array|\Cake\Database\ExpressionInterface $conditions The conditions to test. Must be a ExpressionInterface
+     * @param array|mixed|\Cake\Database\ExpressionInterface $conditions The conditions to test. Must be a ExpressionInterface
```
Comment (dereuromark) [inline src/Database/Expression/CaseExpression.php]: Why mixed? That kind of allows everything now.

File: tests/TestCase/Database/Expression/CaseExpressionTest.php [hunk 5]
```diff
@@ -53,6 +54,40 @@ public function testSqlOutput()
...
+            ->then('testResult');
+        $expected = 'CASE testColumn WHEN :param0 THEN :param1 END';
+        $this->assertSame($expected, $caseExpression->sql(new ValueBinder()));
+
+        $caseExpression = new CaseExpression();
+        $caseExpression
+            ->when('test1')
+            ->then('test2')
+            ->then('test4')
+            ->when('test3')
```
Comment (markstory) [inline tests/TestCase/Database/Expression/CaseExpressionTest.php]: Why are there two then() calls in a row?

File: tests/TestCase/Database/Expression/CaseExpressionTest.php [hunk 6]
```diff
@@ -53,6 +54,40 @@ public function testSqlOutput()
...
+            ->elseValue('testElse')
+            ->caseValue('test0');
+        $expected = 'CASE :param0 WHEN :param1 THEN :param2 WHEN :param3 THEN :param4 ELSE :param5 END';
+        $this->assertSame($expected, $caseExpression->sql(new ValueBinder()));
```
Comment (markstory) [inline tests/TestCase/Database/Expression/CaseExpressionTest.php]: Should there also be a test for improper usage creating invalid SQL?

File: src/Database/Expression/CaseExpression.php (hunk 4 above)
Comment (raul338) [reply src/Database/Expression/CaseExpression.php]: Sorry, fixed and updated

**Comment by markstory** [comment]
> There is also the issue if move $query->newCaseExpr() to something else , perhaps $query->newExpr()->newCase()

I would prefer case expressions to hang off QueryExpression like all the other expression sub types. Right now it is an odd ball being on Query.

File: tests/TestCase/Database/Expression/CaseExpressionTest.php (hunk 5 above)
Comment (raul338) [reply tests/TestCase/Database/Expression/CaseExpressionTest.php]: I test 2 when/then, as long as they're balanced it should work

Comment (markstory) [reply tests/TestCase/Database/Expression/CaseExpressionTest.php]: Ok.

Comment (lorenzo) [reply tests/TestCase/Database/Expression/CaseExpressionTest.php]: Was this change merged recently?
Can we please avoid calling it multiple times? I thought we would have a different `MultiCaseExpression` for multiple cases with different syntax. It makes not sense the way it is now

Comment (raul338) [reply tests/TestCase/Database/Expression/CaseExpressionTest.php]: This PR introduces this change. I'm not a fan of the when/then calls either (I prefer them in a single call via `add`)

What would be different in the `MultiCaseExpression` than this CaseExpression ? (never saw that in the original issue). Shouldn't both handle a single case with multiple when/then sub-clauses?

**Comment by raul338** [comment]
How should I continue with this. Should I drop the then/when calls? Or adding more tests only?

**Comment by lorenzo** [comment]
@raul338 let me think a bit more about what a good API would be. In the mean time, I think dropping the `then/when` calls is the way to go

**Comment by raul338** [comment]
In the current state the API would be like this

```php
<?php
// Deprecated
$query
    ->newExpr()
    ->addCase(
        [
            // Conditions
            $query
                ->newExpr()
                ->add(['published' => 'Y']),
            $query
                ->newExpr()
                ->add(['published' => 'N'])
        ],
        [
            // Values
            'Published',
            'Not published',
            'None' // Else value
        ]
    );

// Simple Case
$query
    ->newExpr()
    ->addCase($query->identifier('published'))
        ->add('Y', 'Published')
        ->add('N', 'Not Published')
        ->elseValue('None');

$query
    ->newExpr()
    ->addCase()
        ->caseValue('published')
        ->add('Y', 'Published')
        ->add('N', 'Not Published')
        ->elseValue('None');

// Searched Case
$query
    ->newExpr()
    ->addCase()
        ->add(
            $query->newExpr()->eq('published', 'Y'),
            'Published'
        )
        ->add(
            $query->newExpr()->eq('published', 'N'),
            'Not Published'
        )
        ->elseValue('None');
```

If this is OK, I think just some tests would be needed to merge this PR

File: src/Database/Expression/CaseExpression.php [hunk 7]
```diff
@@ -246,4 +281,26 @@ public function traverse(callable $visitor)
...
+    /**
+     * Helper build a fluent case builder
+     * If Value is not an ExpressionInterface, it will be treated as a value
+     *
+     * @param string $partName property to append the part
+     * @param string|\Cake\Database\ExpressionInterface $value Boolean Condition or Case Value
+     * @param string $type value type
+     * @return $this
+     */
+    protected function _addPart($partName, $value, $type = null)
```
Comment (dereuromark) [inline src/Database/Expression/CaseExpression.php]: Is this method still used?

Comment (raul338) [reply src/Database/Expression/CaseExpression.php]: Sorry, not anymore, I'll work in this PR maybe on weekend.

**Pull request closed without merging**